File: csharp/1377.frog-position-after-t-seconds.cs
/*
 * @lc app=leetcode id=1377 lang=csharp
 *
 * [1377] Frog Position After T Seconds
 */

using System.Collections.Generic;

// @lc code=start
public class Solution {
    public double FrogPosition(int n, int[][] edges, int t, int target) {
        if (n == 1) {
            return 1.0;
        }
        double p = -1;
        int depth = -1;
        bool hasChild = false;
        var queue = new Queue<(double Probability, int Index)>();
        var map = new Dictionary<int, HashSet<int>>();
        foreach (var edge in edges) {
            int a = edge[0], b = edge[1];
            if (!map.TryGetValue(a, out var ha)) {
                ha = new HashSet<int>();
                map[a] = ha;
            }
            if (!map.TryGetValue(b, out var hb)) {
                hb = new HashSet<int>();
                map[b] = hb; 
            }
            ha.Add(b);
            hb.Add(a);
        }
        queue.Enqueue((Probability: 1.0, Index: 1));
        foreach (int next in map[1]) {
            map[next].Remove(1);
        }
        while (queue.Count > 0 && p < 0) {
            ++depth;
            for (int i = queue.Count; i > 0; --i) {
                var curr = queue.Dequeue();
                double np = map[curr.Index].Count > 0 ? 1.0 / (double)map[curr.Index].Count * curr.Probability : -1;
                if (curr.Index == target) {
                    p = curr.Probability;
                    hasChild = np >= 0;
                    break;
                }
                foreach (int next in map[curr.Index]) {
                    queue.Enqueue((Probability: np, Index: next));
                    map[next].Remove(curr.Index);
                }
            }
        }
        if (t < depth) {
            return 0;
        }
        if (t > depth) {
            return hasChild ? 0 : p;
        }
        return p;
    }
}
// @lc code=end

File: java/1377.frog-position-after-t-seconds.java
/*
 * @lc app=leetcode id=1377 lang=java
 *
 * [1377] Frog Position After T Seconds
 */

import java.util.ArrayDeque;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Queue;
import java.util.Set;

// @lc code=start
class Solution {
    public double frogPosition(int n, int[][] edges, int t, int target) {
        if (n == 1) {
            return 1.0;
        }
        Map<Integer, Set<Integer>> neighbors = new HashMap<>(n);
        Queue<Pair> queue = new ArrayDeque<>();
        --target;
        for (int[] edge: edges) {
            int l = edge[0] - 1, r = edge[1] - 1;
            neighbors.computeIfAbsent(l, ignored -> new HashSet<>()).add(r);
            neighbors.computeIfAbsent(r, ignored -> new HashSet<>()).add(l);
        }
        for (int nb : neighbors.get(0)) {
            neighbors.get(nb).remove(0);
        }
        queue.offer(new Pair(1.0, 0));
        int depth = 0;
        double p = -1;
        while (!queue.isEmpty() && p < 0) {
            for (int k = queue.size(); k > 0; --k) {
                Pair curr = queue.poll();
                if (curr.i() == target) {
                    if (t < depth || (t > depth && !neighbors.get(target).isEmpty())) {
                        return 0;
                    }
                    return curr.p();
                }
                var nbs = neighbors.get(curr.i());
                if (nbs.isEmpty()) {
                    continue;
                }
                double np = 1.0 / (double)nbs.size() * curr.p();
                for (int nb : nbs) {
                    queue.offer(new Pair(np, nb));
                    var nbnbs = neighbors.get(nb);
                    nbnbs.remove(curr.i());
                }
            }
            ++depth;
        }
        throw new IllegalStateException();
    }

    private record Pair(double p, int i) {}
}
// @lc code=end

